### DISCUSSION TITLE
Post-parsing middleware / coercing explicit nulls in requests to default values

### DISCUSSION THREAD

**Discussion by acdha** [fastapi/fastapi#11457] [Questions] [answered]
### First Check

- [X] I added a very descriptive title here.
- [X] I used the GitHub search to find a similar question and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/pydantic/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
# Basically the model example except that the default for tax is not None

from fastapi import FastAPI
from pydantic import BaseModel


class Item(BaseModel):
    name: str
    description: str | None = None
    price: float
    tax: float | None = Field(default=1.0)


app = FastAPI()


@app.post("/items/")
async def create_item(item: Item):
    return item
```

### Description

I have an app which has a number of fields defined as `my_field: SomeType | None = Field(default=sensible_value)` because the original intention was to accept requests like `{"my_field": "some value"}` or `{}` (getting the default), but not `{`"my_field": null`}`. 

I would like to change all of those to remove the `| None` but I would like to non-destructively determine whether there are in fact any clients making such requests, which means I'd like a transition period running some logic along the lines of “if the value is None and there's a default value for that field, use the default instead and log the request for the deprecated usage”. 

Obviously I can do that with some model validators but it seems like there should be a way to do this without needing to touch each model. Middleware would be an obvious choice but that happens earlier in the request process and what I'd really like would be to have it load as soon as the body has been parsed into a model. I mention the case of explicitly mapping nulls to default values but there are variations on this idea where we have some similar common parameters used across models where the same kind of post-parse middleware would be useful as a standard mechanism rather than something which has to be added to each model.

### Operating System

Linux, macOS

### FastAPI Version

0.109.2

### Pydantic Version

2.7.0

### Python Version

3.12

**Comment by YuriiMotov** [comment, marked as answer]
IMO, the easiest way to remove nulls in input data is to implement model_validator for base model and use this base model for all your models:
```

from typing import Any

from pydantic import BaseModel, Field, model_validator


class MyBaseModel(BaseModel):

    @model_validator(mode="before")
    @classmethod
    def remove_nulls(cls, data: dict[str, Any] | Any) -> dict[str, Any] | Any:
        if not isinstance(data, dict):
            return data
        for key in list(data.keys()):
            if data[key] is None:
                data.pop(key)
        return data


class Item(MyBaseModel):
    name: str
    description: str | None = None
    price: float
    tax: float | None = Field(default=1.0)


i1 = Item.model_validate(
    {
        "name": "n",
        "description": None,
        "price": 1.2,
        "tax": None
    }
)

assert i1.description is None
assert i1.tax == 1.0

i2 = Item.model_validate(
    {
        "name": "n",
        "price": 1.2,
    }
)
assert i2.description is None
assert i2.tax == 1.0
```

**Reply by acdha** [reply]
Thanks, the subclassing approach turned out to be less invasive than I'd been worried about.
